## INFORME DE ANÁLISIS EXPLORATORIO DE DATOS
### BMW Pricing Dataset

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import LabelEncoder,MinMaxScaler

In [3]:
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)

### 1. Análisis inicial

In [3]:
# Importamos los datos a un DataFrame de pandas
bmw = pd.read_csv('dataset/bmw_pricing_v3.csv')

In [4]:
bmw.shape

(4843, 18)

In [5]:
bmw.sample(5)

marca modelo        km  potencia fecha_registro tipo_gasolina   color  \
822    NaN    520  191841.0     135.0     2014-04-01        diesel   black   
1993   NaN    116  146057.0      85.0     2012-06-01        diesel     NaN   
4559   BMW     X1  134886.0     120.0            NaN        diesel  silver   
182    BMW    520  146528.0     140.0     2014-07-01        diesel     NaN   
233    BMW    520  173446.0     135.0     2013-02-01        diesel   black   

     tipo_coche volante_regulable aire_acondicionado camara_trasera  \
822         NaN              True               True           True   
1993        NaN             False               True          False   
4559        suv              True               True          False   
182         NaN              True               True           True   
233      estate              True               True          False   

     asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
822                          NaN                 True      True  True   
1993                         NaN                False     False  True   
4559                       False                 True     False  True   
182                          NaN                False     False  True   
233                          NaN                 True     False  True   

     alerta_lim_velocidad   precio fecha_venta  
822                  True  19300.0  2018-03-01  
1993                False   8900.0  2018-03-01  
4559                 True  15200.0  2018-07-01  
182                   NaN  18600.0  2018-08-01  
233                 False  15600.0  2018-01-01

In [6]:
bmw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   marca                        3873 non-null   object 
 1   modelo                       4840 non-null   object 
 2   km                           4841 non-null   float64
 3   potencia                     4842 non-null   float64
 4   fecha_registro               2420 non-null   object 
 5   tipo_gasolina                4838 non-null   object 
 6   color                        4398 non-null   object 
 7   tipo_coche                   3383 non-null   object 
 8   volante_regulable            4839 non-null   object 
 9   aire_acondicionado           4357 non-null   object 
 10  camara_trasera               4841 non-null   object 
 11  asientos_traseros_plegables  1452 non-null   object 
 12  elevalunas_electrico         4841 non-null   object 
 13  bluetooth         

In [7]:
bmw.describe()

km     potencia         precio
count  4.841000e+03  4842.000000    4837.000000
mean   1.409593e+05   128.981826   15831.920612
std    6.020853e+04    38.994839    9222.630708
min   -6.400000e+01     0.000000     100.000000
25%    1.028840e+05   100.000000   10800.000000
50%    1.410800e+05   120.000000   14200.000000
75%    1.752170e+05   135.000000   18600.000000
max    1.000376e+06   423.000000  178500.000000

In [10]:
bmw.describe(include=[object])

marca modelo fecha_registro tipo_gasolina  color tipo_coche  \
count   3873   4840           2420          4838   4398       3383   
unique     1     76            180             5     10          8   
top      BMW    320     2013-07-01        diesel  black     estate   
freq    3873    752             93          4631   1499       1107   

       volante_regulable aire_acondicionado camara_trasera  \
count               4839               4357           4841   
unique                 2                  2              2   
top                 True               True          False   
freq                2662               3465           3864   

       asientos_traseros_plegables elevalunas_electrico bluetooth  \
count                         1452                 4841      4115   
unique                           2                    2         2   
top                          False                False     False   
freq                          1150                 2612      3119   

       alerta_lim_velocidad fecha_venta  
count                  4115        4842  
unique                    2          13  
top                    True  2018-05-01  
freq                   2240         809

### 2. Análisis de duplicados

In [11]:
bmw.duplicated().value_counts()

False    4843
Name: count, dtype: int64

In [13]:
bmw[bmw.duplicated()]

Empty DataFrame
Columns: [marca, modelo, km, potencia, fecha_registro, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, asientos_traseros_plegables, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

No hay registros duplicados

### 3. Tratamiento de datos nulos

In [14]:
bmw.isnull().sum()

marca                           970
modelo                            3
km                                2
potencia                          1
fecha_registro                 2423
tipo_gasolina                     5
color                           445
tipo_coche                     1460
volante_regulable                 4
aire_acondicionado              486
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

- 3.1- MARCA

In [16]:
bmw.marca.unique()

array([nan, 'BMW'], dtype=object)

In [18]:
round(bmw.marca.value_counts(dropna=False, normalize=True) * 100, 2)

marca
BMW    79.97
NaN    20.03
Name: proportion, dtype: float64

In [19]:
# Copia de seguridad
bmw2 = bmw.copy()

Esta columna no aporta información y contiene muchos nulos. Se elimina

In [20]:
# Eliminamos la columna
del(bmw2['marca'])

In [21]:
bmw2.head()

modelo        km  potencia fecha_registro tipo_gasolina   color  \
0    118  140411.0     100.0     2012-02-01        diesel   black   
1     M4   13929.0     317.0            NaN        petrol    grey   
2    320  183297.0     120.0     2012-04-01        diesel   white   
3    420  128035.0     135.0            NaN        diesel     red   
4    425   97097.0     160.0            NaN        diesel  silver   

    tipo_coche volante_regulable aire_acondicionado camara_trasera  \
0          NaN              True               True          False   
1  convertible              True               True          False   
2          NaN             False              False          False   
3  convertible              True               True          False   
4          NaN              True               True          False   

  asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
0                         NaN                 True       NaN  True   
1                         NaN                False      True  True   
2                         NaN                 True     False  True   
3                         NaN                 True      True  True   
4                       False                False      True  True   

  alerta_lim_velocidad   precio fecha_venta  
0                  NaN  11300.0  2018-01-01  
1                 True  69700.0  2018-02-01  
2                False  10200.0  2018-02-01  
3                  NaN  25100.0  2018-02-01  
4                 True  33400.0  2018-04-01

- 3.2- FECHA_REGISTRO

In [22]:
round(bmw2.fecha_registro.value_counts(dropna=False, normalize=True) * 100, 2)

fecha_registro
NaN           50.03
2013-07-01     1.92
2014-03-01     1.71
2014-05-01     1.67
2013-01-01     1.61
2013-10-01     1.59
2014-02-01     1.53
2013-06-01     1.38
2014-01-01     1.38
2013-02-01     1.38
2013-11-01     1.26
2012-10-01     1.26
2013-04-01     1.22
2014-04-01     1.20
2013-09-01     1.20
2013-03-01     1.20
2013-08-01     1.12
2012-06-01     1.12
2014-07-01     1.12
2014-09-01     1.01
2013-05-01     0.95
2014-08-01     0.93
2012-11-01     0.89
2012-03-01     0.87
2014-06-01     0.85
2013-12-01     0.83
2012-01-01     0.72
2014-10-01     0.72
2014-11-01     0.70
2012-04-01     0.70
2012-07-01     0.68
2012-09-01     0.66
2012-08-01     0.60
2015-03-01     0.58
2012-05-01     0.56
2012-12-01     0.54
2012-02-01     0.39
2015-06-01     0.39
2015-10-01     0.39
2015-01-01     0.37
2015-05-01     0.37
2014-12-01     0.29
2015-04-01     0.29
2011-10-01     0.29
2011-06-01     0.29
2011-05-01     0.25
2015-07-01     0.25
2011-12-01     0.25
2011-09-01     0.23
2011-

La columna contiene demasiados nulos. Se elimina

In [23]:
bmw3 = bmw2.copy()

In [24]:
del(bmw3['fecha_registro'])

In [25]:
bmw3.head()

modelo        km  potencia tipo_gasolina   color   tipo_coche  \
0    118  140411.0     100.0        diesel   black          NaN   
1     M4   13929.0     317.0        petrol    grey  convertible   
2    320  183297.0     120.0        diesel   white          NaN   
3    420  128035.0     135.0        diesel     red  convertible   
4    425   97097.0     160.0        diesel  silver          NaN   

  volante_regulable aire_acondicionado camara_trasera  \
0              True               True          False   
1              True               True          False   
2             False              False          False   
3              True               True          False   
4              True               True          False   

  asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
0                         NaN                 True       NaN  True   
1                         NaN                False      True  True   
2                         NaN                 True     False  True   
3                         NaN                 True      True  True   
4                       False                False      True  True   

  alerta_lim_velocidad   precio fecha_venta  
0                  NaN  11300.0  2018-01-01  
1                 True  69700.0  2018-02-01  
2                False  10200.0  2018-02-01  
3                  NaN  25100.0  2018-02-01  
4                 True  33400.0  2018-04-01

In [26]:
bmw3.isnull().sum()

modelo                            3
km                                2
potencia                          1
tipo_gasolina                     5
color                           445
tipo_coche                     1460
volante_regulable                 4
aire_acondicionado              486
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

- 3.3- COLOR

In [27]:
round(bmw3.color.value_counts(dropna=False, normalize=True) * 100, 2)

color
black     30.95
grey      22.09
blue      13.36
white      9.99
NaN        9.19
brown      6.24
silver     6.03
red        0.97
beige      0.76
green      0.29
orange     0.12
Name: proportion, dtype: float64

9,19% de nulos de la muestra total. No se eliminan, sino que se imputa el valor 'otros'

In [28]:
bmw4 = bmw3.copy()

In [29]:
# Se imputa a los nulos el valor 'otros'
bmw4['color'].fillna('otros', inplace=True)

In [30]:
round(bmw4['color'].value_counts(dropna=False, normalize=True) * 100, 2)

color
black     30.95
grey      22.09
blue      13.36
white      9.99
otros      9.19
brown      6.24
silver     6.03
red        0.97
beige      0.76
green      0.29
orange     0.12
Name: proportion, dtype: float64

In [31]:
bmw4.isnull().sum()

modelo                            3
km                                2
potencia                          1
tipo_gasolina                     5
color                             0
tipo_coche                     1460
volante_regulable                 4
aire_acondicionado              486
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

- 3.4- TIPO_COCHE

In [32]:
round(bmw4.tipo_coche.value_counts(dropna=False, normalize=True) * 100, 2)

tipo_coche
NaN            30.15
estate         22.86
sedan          16.95
suv            15.59
hatchback      10.08
subcompact      1.59
coupe           1.57
convertible     0.62
van             0.60
Name: proportion, dtype: float64

Hay un 30,15% de nulos, pero la columna es significativa. Se intentará imputar valores a los registros nulos

In [33]:
bmw4.tipo_coche.value_counts()

tipo_coche
estate         1107
sedan           821
suv             755
hatchback       488
subcompact       77
coupe            76
convertible      30
van              29
Name: count, dtype: int64

In [50]:
bmw4.groupby(['modelo', 'potencia', 'tipo_coche'], dropna=False)[['modelo']].count()

modelo
modelo            potencia tipo_coche         
 Active Tourer    100.0    van               1
114               70.0     hatchback         3
                           subcompact        4
                           NaN               3
                  75.0     hatchback         3
                           NaN               2
116               80.0     estate            1
                           hatchback         2
                           subcompact        1
                  85.0     estate            1
                           hatchback       182
                           sedan             3
                           subcompact       55
                           NaN             104
                  87.0     hatchback         1
                  90.0     hatchback         1
                           NaN               1
                  100.0    hatchback         3
                           NaN               1
                  101.0    hatchback         1
                  105.0    hatchback         1
118               90.0     hatchback         8
                           subcompact        3
                           NaN               5
                  95.0     subcompact        1
                           NaN               1
                  100.0    coupe             1
                           hatchback        16
                           subcompact        3
                           NaN              12
                  105.0    convertible       1
                           coupe             1
                           hatchback        37
                           sedan             3
                           subcompact        5
                           NaN              19
                  110.0    hatchback        17
                           NaN               7
                  125.0    subcompact        1
                           NaN               1
                  127.0    hatchback         1
120               110.0    hatchback         1
                  120.0    convertible       2
                           subcompact        1
                           NaN               1
                  130.0    convertible       2
                           hatchback         2
                           NaN               2
                  135.0    hatchback        11
                           sedan             1
                           subcompact        1
                           NaN               9
                  140.0    hatchback         2
                  142.0    NaN               1
123               150.0    hatchback         1
                           NaN               1
125               155.0    hatchback         1
                  160.0    hatchback         4
135               225.0    convertible       1
                           coupe             1
                  235.0    hatchback         2
214 Gran Tourer   70.0     van               1
216               85.0     van               1
216 Active Tourer 85.0     van               3
216 Gran Tourer   85.0     van               1
                           NaN               3
218               100.0    convertible       2
                           coupe             1
                           NaN               1
                  105.0    coupe             1
                           NaN               1
                  110.0    convertible       1
                           van               2
                           NaN               3
218 Active Tourer 100.0    van               4
                           NaN               2
                  110.0    van               4
                           NaN               1
                  112.0    van               2
218 Gran Tourer   85.0     NaN               1
                  100.0    van               1
                  110.0    van               4
                           NaN               3
220               135.0    convertible       

Hay demasiada variabilidad para imputar valores manualmente. Se le imputa el valor 'otros'

In [51]:
bmw5 = bmw4.copy()

In [52]:
bmw5.tipo_coche.fillna('otros', inplace=True)

In [53]:
round(bmw5.tipo_coche.value_counts(dropna=False, normalize=True) * 100, 2)

tipo_coche
otros          30.15
estate         22.86
sedan          16.95
suv            15.59
hatchback      10.08
subcompact      1.59
coupe           1.57
convertible     0.62
van             0.60
Name: proportion, dtype: float64

In [54]:
bmw5.isnull().sum()

modelo                            3
km                                2
potencia                          1
tipo_gasolina                     5
color                             0
tipo_coche                        0
volante_regulable                 4
aire_acondicionado              486
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

- 3.5- AIRE ACONDICIONADO

In [55]:
round(bmw5.aire_acondicionado.value_counts(dropna=False, normalize=True) * 100, 2)

aire_acondicionado
True     71.55
False    18.42
NaN      10.04
Name: proportion, dtype: float64

Hay un 10% de nulos. Se imputa el valor 'sin_info'

In [56]:
bmw6 = bmw5.copy()

In [57]:
bmw6.aire_acondicionado.fillna('sin_info', inplace=True)

In [58]:
round(bmw6.aire_acondicionado.value_counts(dropna=False,normalize=True)*100, 2)

aire_acondicionado
True        71.55
False       18.42
sin_info    10.04
Name: proportion, dtype: float64

In [59]:
bmw6.isnull().sum()

modelo                            3
km                                2
potencia                          1
tipo_gasolina                     5
color                             0
tipo_coche                        0
volante_regulable                 4
aire_acondicionado                0
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

- 3.6- ASIENTOS_TRASEROS_PLEGABLES

In [60]:
round(bmw6.asientos_traseros_plegables.value_counts(dropna=False,normalize=True)*100, 2)

asientos_traseros_plegables
NaN      70.02
False    23.75
True      6.24
Name: proportion, dtype: float64

Hay demasiados nulos. Se elimina la columna

In [61]:
bmw7 = bmw6.copy()

In [62]:
del(bmw7['asientos_traseros_plegables'])

In [63]:
bmw7.head()

modelo        km  potencia tipo_gasolina   color   tipo_coche  \
0    118  140411.0     100.0        diesel   black        otros   
1     M4   13929.0     317.0        petrol    grey  convertible   
2    320  183297.0     120.0        diesel   white        otros   
3    420  128035.0     135.0        diesel     red  convertible   
4    425   97097.0     160.0        diesel  silver        otros   

  volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
0              True               True          False                 True   
1              True               True          False                False   
2             False              False          False                 True   
3              True               True          False                 True   
4              True               True          False                False   

  bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
0       NaN  True                  NaN  11300.0  2018-01-01  
1      True  True                 True  69700.0  2018-02-01  
2     False  True                False  10200.0  2018-02-01  
3      True  True                  NaN  25100.0  2018-02-01  
4      True  True                 True  33400.0  2018-04-01

In [64]:
bmw7.isnull().sum()

modelo                    3
km                        2
potencia                  1
tipo_gasolina             5
color                     0
tipo_coche                0
volante_regulable         4
aire_acondicionado        0
camara_trasera            2
elevalunas_electrico      2
bluetooth               728
gps                       0
alerta_lim_velocidad    728
precio                    6
fecha_venta               1
dtype: int64

- 3.7- BLUETOOTH

In [65]:
round(bmw7.bluetooth.value_counts(dropna=False,normalize=True)*100, 2)

bluetooth
False    64.40
True     20.57
NaN      15.03
Name: proportion, dtype: float64

Hay un 15% de nulos. Se imputa el valor 'sin_info'

In [68]:
bmw8 = bmw7.copy()

In [69]:
bmw8.bluetooth.fillna('sin_info', inplace=True)

In [70]:
round(bmw8.bluetooth.value_counts(dropna=False, normalize=True) * 100, 2)

bluetooth
False       64.40
True        20.57
sin_info    15.03
Name: proportion, dtype: float64

In [71]:
bmw8.isnull().sum()

modelo                    3
km                        2
potencia                  1
tipo_gasolina             5
color                     0
tipo_coche                0
volante_regulable         4
aire_acondicionado        0
camara_trasera            2
elevalunas_electrico      2
bluetooth                 0
gps                       0
alerta_lim_velocidad    728
precio                    6
fecha_venta               1
dtype: int64

- 3.8- ALERTA_LIM_VELOCIDAD

In [72]:
round(bmw8.alerta_lim_velocidad.value_counts(dropna=False,normalize=True)*100, 2)

alerta_lim_velocidad
True     46.25
False    38.72
NaN      15.03
Name: proportion, dtype: float64

Hay un 15% de nulos. Se imputa el valor 'sin_info'

In [73]:
bmw9 = bmw8.copy()

In [74]:
bmw9.alerta_lim_velocidad.fillna('sin_info', inplace=True)

In [75]:
round(bmw9.alerta_lim_velocidad.value_counts(dropna=False, normalize=True) * 100, 2)

alerta_lim_velocidad
True        46.25
False       38.72
sin_info    15.03
Name: proportion, dtype: float64

In [76]:
bmw9.isnull().sum()

modelo                  3
km                      2
potencia                1
tipo_gasolina           5
color                   0
tipo_coche              0
volante_regulable       4
aire_acondicionado      0
camara_trasera          2
elevalunas_electrico    2
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

3.9- Categorías con pocos nulos

- 3.9.1- MODELO

In [77]:
bmw9[bmw9.modelo.isnull()]

modelo        km  potencia tipo_gasolina   color tipo_coche  \
173     NaN  146338.0     105.0        diesel   black      otros   
4766    NaN  115566.0     105.0        diesel  silver        suv   
4802    NaN  118640.0     110.0        diesel   white      otros   

     volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
173              False               True          False                False   
4766             False               True          False                 True   
4802             False              False          False                False   

     bluetooth    gps alerta_lim_velocidad   precio fecha_venta  
173   sin_info   True             sin_info  13300.0  2018-08-01  
4766     False   True                False  14900.0  2018-09-01  
4802     False  False                False    800.0  2018-03-01

Se evalúan los 3 registros nulos de forma individual

In [78]:
# Vehículo índice 173
round(bmw9[(bmw9['potencia']==105) & (bmw9['tipo_gasolina']=='diesel')&(bmw9['aire_acondicionado']==True)&(bmw9['camara_trasera']==False)&(bmw9['elevalunas_electrico']==False)&(bmw9['volante_regulable']==False)&(bmw9['gps']==True)]['modelo'].value_counts(normalize=True)*100, 2)

modelo
318                 63.11
X1                  11.65
118                 10.68
318 Gran Turismo     9.71
X3                   3.88
525                  0.97
Name: proportion, dtype: float64

In [79]:
# Se imputa el modelo 318 por probabilidad
bmw9.loc[173, 'modelo'] = '318'

In [81]:
# Comprobación
bmw9.loc[[173]]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
173    318  146338.0     105.0        diesel  black      otros   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
173             False               True          False                False   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
173  sin_info  True             sin_info  13300.0  2018-08-01

In [82]:
bmw9[bmw9.modelo.isnull()]

modelo        km  potencia tipo_gasolina   color tipo_coche  \
4766    NaN  115566.0     105.0        diesel  silver        suv   
4802    NaN  118640.0     110.0        diesel   white      otros   

     volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
4766             False               True          False                 True   
4802             False              False          False                False   

     bluetooth    gps alerta_lim_velocidad   precio fecha_venta  
4766     False   True                False  14900.0  2018-09-01  
4802     False  False                False    800.0  2018-03-01

In [ ]:
#Vehículo indice 4766
bmw9[(bmw9['potencia']==105) & (bmw9['tipo_gasolina']=='diesel')&(bmw9['aire_acondicionado']==True)&(bmw9['camara_trasera']==False)&(bmw9['elevalunas_electrico']==True)&(bmw9['volante_regulable']==False)&(bmw9['gps']==True)&(bmw9['tipo_coche']=='suv')&(bmw9['bluetooth']==False)]['modelo'].value_counts(normalize=True)*100

modelo
X1    62.5
X3    37.5
Name: proportion, dtype: float64

In [83]:
# Se imputa el modelo X1 por probabilidad
bmw9.loc[4766, 'modelo'] = 'X1'

In [84]:
# Comprobación
bmw9.loc[[4766]]

modelo        km  potencia tipo_gasolina   color tipo_coche  \
4766     X1  115566.0     105.0        diesel  silver        suv   

     volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
4766             False               True          False                 True   

     bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
4766     False  True                False  14900.0  2018-09-01

In [85]:
bmw9[bmw9.modelo.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
4802    NaN  118640.0     110.0        diesel  white      otros   

     volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
4802             False              False          False                False   

     bluetooth    gps alerta_lim_velocidad  precio fecha_venta  
4802     False  False                False   800.0  2018-03-01

In [86]:
# Vehículo índice 4802
round(bmw9[(bmw9['potencia']==110) & (bmw9['tipo_gasolina']=='diesel')&(bmw9['aire_acondicionado']==False)&(bmw9['camara_trasera']==False)&(bmw9['elevalunas_electrico']==False)&(bmw9['volante_regulable']==False)&(bmw9['gps']==False)&(bmw9['bluetooth']==False)]['modelo'].value_counts(normalize=True)*100, 2)

modelo
X3     33.33
218    33.33
318    16.67
X1     16.67
Name: proportion, dtype: float64

Dado que no hay un modelo con mayor probabilidad, se elimina el registro

In [87]:
bmw9.drop(index=4802, inplace=True)

In [88]:
bmw9[bmw9.modelo.isnull()]

Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

In [89]:
bmw9.isnull().sum()

modelo                  0
km                      2
potencia                1
tipo_gasolina           5
color                   0
tipo_coche              0
volante_regulable       4
aire_acondicionado      0
camara_trasera          2
elevalunas_electrico    2
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

- 3.9.2- KM

In [90]:
bmw9[bmw9.km.isnull()]

modelo  km  potencia tipo_gasolina  color tipo_coche volante_regulable  \
459     520 NaN     135.0        diesel  black      otros              True   
4811     X3 NaN     190.0        diesel   blue        van              True   

     aire_acondicionado camara_trasera elevalunas_electrico bluetooth   gps  \
459                True           True                 True      True  True   
4811               True           True                 True     False  True   

     alerta_lim_velocidad   precio fecha_venta  
459                  True  20300.0  2018-02-01  
4811                 True  19500.0  2018-07-01

No hay manera de deducir unos valores fiables para imputar a los registros nulos. Por lo tanto, se eliminan

In [91]:
bmw9.dropna(subset='km', inplace=True)

In [92]:
bmw9[bmw9.km.isnull()]

Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

In [93]:
bmw9.isnull().sum()

modelo                  0
km                      0
potencia                1
tipo_gasolina           5
color                   0
tipo_coche              0
volante_regulable       4
aire_acondicionado      0
camara_trasera          2
elevalunas_electrico    2
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

- 3.9.3- POTENCIA

In [94]:
bmw9[bmw9.potencia.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
181    525  106054.0       NaN        diesel  black     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
181             False               True          False                False   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
181     False  True                 True  18700.0  2018-03-01

In [108]:
#Podemos deducir la potencia segun modelo y otras variables?
modelo_181 = '525'
tipo_gasolina_181 = 'diesel'
tipo_coche_181 = 'estate'
volante_regulable_181 = False
aire_acondicionado_181 = True
elevalunas_electrico_181 = False
filtro = (bmw9.modelo == modelo_181) & (bmw9.tipo_gasolina == tipo_gasolina_181) & (bmw9.tipo_coche == tipo_coche_181) & (bmw9.volante_regulable == volante_regulable_181) & (bmw9.aire_acondicionado == aire_acondicionado_181) & (bmw9.elevalunas_electrico == elevalunas_electrico_181)
bmw9[['potencia', 'modelo', 'tipo_gasolina', 'tipo_coche', 'volante_regulable', 'aire_acondicionado', 'elevalunas_electrico']][filtro].value_counts().reset_index()

potencia modelo tipo_gasolina tipo_coche  volante_regulable  \
0     160.0    525        diesel     estate              False   

   aire_acondicionado  elevalunas_electrico  count  
0                True                 False      4

In [109]:
# Se imputa el valor
bmw9.loc[181, 'potencia'] = 160

In [110]:
# Comprobación
bmw9.loc[[181]]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
181    525  106054.0     160.0        diesel  black     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
181             False               True          False                False   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
181     False  True                 True  18700.0  2018-03-01

In [111]:
bmw9[bmw9.potencia.isnull()]

Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

In [5]:
bmw9.isnull().sum()

modelo                  0
km                      0
potencia                0
tipo_gasolina           5
color                   0
tipo_coche              0
volante_regulable       4
aire_acondicionado      0
camara_trasera          2
elevalunas_electrico    2
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

- 3.9.4- TIPO_GASOLINA

In [13]:
bmw9[bmw9.tipo_gasolina.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
82     420   54993.0     135.0           NaN  black      otros   
185    320  186697.0     135.0           NaN  white     estate   
444    318  111622.0     100.0           NaN  black     estate   
479    318  170350.0     105.0           NaN  white      otros   
629    318  147558.0     105.0           NaN  white     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
82              False              False          False                False   
185             False              False          False                False   
444              True               True          False                False   
479             False               True          False                False   
629             False           sin_info          False                False   

    bluetooth    gps alerta_lim_velocidad   precio fecha_venta  
82      False  False                False  18400.0  2018-03-01  
185     False  False                 True   2400.0  2018-04-01  
444     False   True                 True  15400.0  2018-02-01  
479     False   True                False  10200.0  2018-02-01  
629     False  False                 True    100.0  2018-03-01

In [20]:
#Evaluamos el modelo 420 (indice 82) para ver si podemos imputar el tipo_gasolina
bmw9[['modelo','potencia','tipo_gasolina']][(bmw9['modelo']=='420')&(bmw9['potencia']==135.0)].value_counts().reset_index()

modelo  potencia tipo_gasolina  count
0    420     135.0        diesel     26
1    420     135.0        petrol      2

In [21]:
# Imputamos el valor 'diesel'
bmw9.loc[[82],['tipo_gasolina']]='diesel'
# Comprobación
bmw9.loc[[82],['tipo_gasolina']]

tipo_gasolina
82        diesel

In [22]:
bmw9[bmw9.tipo_gasolina.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
185    320  186697.0     135.0           NaN  white     estate   
444    318  111622.0     100.0           NaN  black     estate   
479    318  170350.0     105.0           NaN  white      otros   
629    318  147558.0     105.0           NaN  white     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
185             False              False          False                False   
444              True               True          False                False   
479             False               True          False                False   
629             False           sin_info          False                False   

    bluetooth    gps alerta_lim_velocidad   precio fecha_venta  
185     False  False                 True   2400.0  2018-04-01  
444     False   True                 True  15400.0  2018-02-01  
479     False   True                False  10200.0  2018-02-01  
629     False  False                 True    100.0  2018-03-01

In [23]:
#Evaluamos el modelo 320 (indice 185) para ver si podemos imputar el tipo_gasolina
bmw9[['modelo','potencia','tipo_gasolina']][(bmw9['modelo']=='320')&(bmw9['potencia']==135.0)].value_counts().reset_index()

modelo  potencia tipo_gasolina  count
0    320     135.0        diesel    297
1    320     135.0        petrol     13

In [24]:
# Imputamos el valor 'diesel'
bmw9.loc[[185],['tipo_gasolina']]='diesel'
# Comprobación
bmw9.loc[[185],['tipo_gasolina']]

tipo_gasolina
185        diesel

In [25]:
bmw9[bmw9.tipo_gasolina.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
444    318  111622.0     100.0           NaN  black     estate   
479    318  170350.0     105.0           NaN  white      otros   
629    318  147558.0     105.0           NaN  white     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
444              True               True          False                False   
479             False               True          False                False   
629             False           sin_info          False                False   

    bluetooth    gps alerta_lim_velocidad   precio fecha_venta  
444     False   True                 True  15400.0  2018-02-01  
479     False   True                False  10200.0  2018-02-01  
629     False  False                 True    100.0  2018-03-01

In [26]:
#Evaluamos el modelo 318  con potencia 100 (indice 444) para ver si podemos imputar el tipo_gasolina
bmw9[['modelo','potencia','tipo_gasolina']][(bmw9['modelo']=='318')&(bmw9['potencia']==100.0)].value_counts().reset_index()

modelo  potencia tipo_gasolina  count
0    318     100.0        diesel    305
1    318     100.0        petrol      3

In [27]:
# Imputamos el valor 'diesel'
bmw9.loc[[444],['tipo_gasolina']]='diesel'
# Comprobación
bmw9.loc[[444],['tipo_gasolina']]

tipo_gasolina
444        diesel

In [28]:
bmw9[bmw9.tipo_gasolina.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
479    318  170350.0     105.0           NaN  white      otros   
629    318  147558.0     105.0           NaN  white     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
479             False               True          False                False   
629             False           sin_info          False                False   

    bluetooth    gps alerta_lim_velocidad   precio fecha_venta  
479     False   True                False  10200.0  2018-02-01  
629     False  False                 True    100.0  2018-03-01

In [29]:
#Evaluamos los modelos 318 con potencia 105 (indices 479 y 629) para ver si podemos imputar el tipo_gasolina
bmw9[['modelo','potencia','tipo_gasolina']][(bmw9['modelo']=='318')&(bmw9['potencia']==105.0)].value_counts().reset_index()

modelo  potencia tipo_gasolina  count
0    318     105.0        diesel    209
1    318     105.0        petrol      3
2    318     105.0        Diesel      1

In [30]:
# Imputamos el valor 'diesel'
bmw9.loc[[479, 629],['tipo_gasolina']]='diesel'
# Comprobación
bmw9.loc[[479, 629],['tipo_gasolina']]

tipo_gasolina
479        diesel
629        diesel

In [31]:
bmw9[bmw9.tipo_gasolina.isnull()]

Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

In [32]:
bmw9.isnull().sum()

modelo                  0
km                      0
potencia                0
tipo_gasolina           0
color                   0
tipo_coche              0
volante_regulable       4
aire_acondicionado      0
camara_trasera          2
elevalunas_electrico    2
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

- 3.9.5- VOLANTE_REGULABLE

In [33]:
bmw9[bmw9['volante_regulable'].isna()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
66     320   93066.0     135.0        diesel  black      coupe   
273    318  195032.0     100.0        diesel   grey      otros   
488    320  227691.0     135.0        diesel  black      otros   
613    318  201626.0     105.0        diesel  black      otros   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
66                NaN              False          False                 True   
273               NaN           sin_info          False                False   
488               NaN               True          False                False   
613               NaN           sin_info          False                 True   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
66      False  True             sin_info  11900.0  2018-06-01  
273     False  True                False   4900.0  2018-01-01  
488      True  True                False  10500.0  2018-02-01  
613     False  True                 True  11700.0  2018-03-01

In [34]:
#Evaluamos el modelo 320 (indice 66) para ver si podemos imputar el volante_regulable
bmw9[['volante_regulable','tipo_gasolina','modelo','potencia','tipo_coche','camara_trasera']][(bmw9['modelo']=='320')&(bmw9['potencia']==135)&(bmw9['tipo_gasolina']=='diesel')&(bmw9['tipo_coche']=='coupe')&(bmw9['camara_trasera']==False)].value_counts(normalize=True).reset_index()

volante_regulable tipo_gasolina modelo  potencia tipo_coche  \
0              False        diesel    320     135.0      coupe   
1               True        diesel    320     135.0      coupe   

   camara_trasera  proportion  
0           False         0.5  
1           False         0.5

Ya que esta variable no se relaciona con las demás características del vehículo, no se puede imputar un valor fiable. Se eliminan los registros nulos.

In [35]:
bmw9.dropna(subset='volante_regulable', inplace=True)

In [36]:
bmw9[bmw9.volante_regulable.isna()]

Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

In [38]:
bmw9.isnull().sum()

modelo                  0
km                      0
potencia                0
tipo_gasolina           0
color                   0
tipo_coche              0
volante_regulable       0
aire_acondicionado      0
camara_trasera          2
elevalunas_electrico    2
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

- 3.9.6- CAMARA_TRASERA

In [39]:
bmw9[bmw9.camara_trasera.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
291    520  163675.0     135.0        diesel  otros     estate   
409    316  160619.0      85.0        diesel  black     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
291              True               True            NaN                 True   
409             False               True            NaN                 True   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
291  sin_info  True                 True  14900.0  2018-01-01  
409  sin_info  True             sin_info  11700.0  2018-04-01

In [40]:
#Evaluamos el modelo 520 (indice 291) para ver si podemos imputar la camara_trasera
bmw9[['camara_trasera','modelo', 'potencia','tipo_gasolina','tipo_coche']][(bmw9['modelo']=='520')&(bmw9['potencia']==135)&(bmw9['tipo_gasolina']=='diesel')&(bmw9['tipo_coche']=='estate')].value_counts(normalize=True).reset_index()

camara_trasera modelo  potencia tipo_gasolina tipo_coche  proportion
0           False    520     135.0        diesel     estate    0.709677
1            True    520     135.0        diesel     estate    0.290323

In [41]:
# Imputamos el valor False
bmw9.loc[[291],['camara_trasera']]=False
# Comprobación
bmw9.loc[291]['camara_trasera']

False

In [42]:
bmw9[bmw9.camara_trasera.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
409    316  160619.0      85.0        diesel  black     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
409             False               True            NaN                 True   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
409  sin_info  True             sin_info  11700.0  2018-04-01

In [43]:
#Evaluamos el modelo 316 (indice 409) para ver si podemos imputar la camara_trasera
bmw9[['camara_trasera','modelo', 'potencia','tipo_gasolina','tipo_coche']][(bmw9['modelo']=='316')&(bmw9['potencia']==85)&(bmw9['tipo_gasolina']=='diesel')&(bmw9['tipo_coche']=='estate')].value_counts(normalize=True).reset_index()

camara_trasera modelo  potencia tipo_gasolina tipo_coche  proportion
0           False    316      85.0        diesel     estate    0.948718
1            True    316      85.0        diesel     estate    0.051282

In [44]:
# Imputamos el valor False
bmw9.loc[[409],['camara_trasera']]=False
# Comprobación
bmw9.loc[409]['camara_trasera']

False

In [45]:
bmw9[bmw9.camara_trasera.isnull()]

Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

In [46]:
bmw9.isnull().sum()

modelo                  0
km                      0
potencia                0
tipo_gasolina           0
color                   0
tipo_coche              0
volante_regulable       0
aire_acondicionado      0
camara_trasera          0
elevalunas_electrico    2
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

- 3.9.7- ELEVALUNAS_ELECTRICO

In [47]:
bmw9[bmw9.elevalunas_electrico.isnull()]

modelo        km  potencia tipo_gasolina  color tipo_coche  \
63     640   69695.0     230.0        diesel  otros      coupe   
305    535  148738.0     230.0        diesel   grey     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
63               True               True          False                  NaN   
305              True               True           True                  NaN   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
63       True  True                 True  36300.0  2018-02-01  
305     False  True                 True  28800.0  2018-01-01

In [48]:
# Evaluamos el modelo 640 (indice 63) para ver si podemos imputar los elevalunas_electricos
bmw9[['elevalunas_electrico','modelo','potencia','tipo_gasolina','tipo_coche']][(bmw9['modelo']=='640')&(bmw9['potencia']==230)&(bmw9['tipo_gasolina']=='diesel')&(bmw9['tipo_coche']=='coupe')].value_counts(normalize=True).reset_index()

elevalunas_electrico modelo  potencia tipo_gasolina tipo_coche  proportion
0                  True    640     230.0        diesel      coupe         1.0

In [53]:
# Imputamos el valor True
bmw9.loc[[63],['elevalunas_electrico']]=True
# Comprobación
bmw9.loc[63,'elevalunas_electrico']

True

In [50]:
bmw9[bmw9.elevalunas_electrico.isnull()]

modelo        km  potencia tipo_gasolina color tipo_coche  \
305    535  148738.0     230.0        diesel  grey     estate   

    volante_regulable aire_acondicionado camara_trasera elevalunas_electrico  \
305              True               True           True                  NaN   

    bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
305     False  True                 True  28800.0  2018-01-01

In [51]:
# Evaluamos el modelo 535 (indice 305) para ver si podemos imputar los elevalunas_electricos
bmw9[['elevalunas_electrico','modelo','potencia','tipo_gasolina','tipo_coche']][(bmw9['modelo']=='535')&(bmw9['potencia']==230)&(bmw9['tipo_gasolina']=='diesel')&(bmw9['tipo_coche']=='estate')].value_counts(normalize=True).reset_index()

elevalunas_electrico modelo  potencia tipo_gasolina tipo_coche  proportion
0                  True    535     230.0        diesel     estate         0.9
1                 False    535     230.0        diesel     estate         0.1

In [54]:
# Imputamos el valor True
bmw9.loc[[305],['elevalunas_electrico']]=True
# Comprobación
bmw9.loc[305,'elevalunas_electrico']

True

In [55]:
bmw9[bmw9.elevalunas_electrico.isnull()]

Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

In [56]:
bmw9.isnull().sum()

modelo                  0
km                      0
potencia                0
tipo_gasolina           0
color                   0
tipo_coche              0
volante_regulable       0
aire_acondicionado      0
camara_trasera          0
elevalunas_electrico    0
bluetooth               0
gps                     0
alerta_lim_velocidad    0
precio                  6
fecha_venta             1
dtype: int64

- 3.9.8- PRECIO

Ya que es la variable que queremos predecir (Target), eliminamos directamente los registros nulos.

- 3.9.9- FECHA_VENTA

### 4. Análisis Univariable

4.1- Evaluamos si hay variables que se puedan convertir a booleanas

4.2- Listamos las variables según si son numéricas, booleanas o categóricas

4.3- Variables numéricas

- 4.3.1- KM

- 4.3.2- POTENCIA

4.4- Variables booleanas

Al ser booleanas, no hay que hacerles ningún otro tratamiento ni evaluación

4.5- Variables categóricas

- 4.5.1- MODELO

- 4.5.2- TIPO_GASOLINA

- 4.5.3- COLOR

- 4.5.4- TIPO_COCHE

- 4.5.5- AIRE_ACONDICIONADO

- 4.5.6- BLUETOOTH

- 4.5.7- ALERTA_LIM_VELOCIDAD

- 4.5.8- FECHA_VENTA

4.6- Target

- 4.6.1- PRECIO

### 5. Análisis de correlación inicial

### 6. Análisis Variable vs Target

- 6.1- Categóricas

- 6.2- Numéricas

- 6.3- Booleanas

### 7. Transformación de variables categóricas. OHE (One-Hot Encoding)

### 8. Normalización de variables numéricas. MinMaxScaler

### 9. Análisis de correlación final